errors with Go 1.27rc2

**Problem Statement**

Fedora Rawhide is introducing go-1.27 and `xq` doesn't work as expected.

~~~
+ go test -buildmode pie -compiler gc -ldflags ' -X github.com/sibprogrammer/xq/version=1.4.0 -extldflags '\''-Wl,-z,relro -Wl,--as-needed  -Wl,-z,pack-relative-relocs -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-hardened-ld-errors -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1 -specs=/usr/lib/rpm/redhat/redhat-package-notes   '\''' ./...
?   	github.com/sibprogrammer/xq	[no test files]
panic: reflect: call of reflect.Value.Int on zero Value
goroutine 13 [running]:
reflect.Value.Int(...)
	/usr/lib/golang/src/reflect/value.go:1471
github.com/sibprogrammer/xq/internal/utils.FormatJson({0x55fc2e2c9718?, 0x26ff98506720?}, {0x55fc2e2c96d8, 0x26ff9857e900}, {0x55fc2dedefa6, 0x2}, 0x0)
	/builddir/build/BUILD/xq-1.4.0-build/xq-1.4.0/internal/utils/utils.go:446 +0x1107
github.com/sibprogrammer/xq/cmd.processContent({0x55fc2e2c95b8?, 0x26ff984f64e0?}, {0x55fc2e2c96d8, 0x26ff9857e900}, 0x26ff984f0300, 0x0, {0x55fc2dedefa6, 0x2}, 0x0)
	/builddir/build/BUILD/xq-1.4.0-build/xq-1.4.0/cmd/root.go:273 +0xb3
github.com/sibprogrammer/xq/cmd.NewRootCmd.func1.3()
	/builddir/build/BUILD/xq-1.4.0-build/xq-1.4.0/cmd/root.go:122 +0x286
created by github.com/sibprogrammer/xq/cmd.NewRootCmd.func1 in goroutine 8
	/builddir/build/BUILD/xq-1.4.0-build/xq-1.4.0/cmd/root.go:111 +0x75f
FAIL	github.com/sibprogrammer/xq/cmd	0.010s
--- FAIL: TestXmlToJSON (0.00s)
panic: reflect: call of reflect.Value.Int on zero Value [recovered, repanicked]
goroutine 9 [running]:
testing.tRunner.func1.2({0x55858618d050, 0x1e64514823a8})
	/usr/lib/golang/src/testing/testing.go:2128 +0x232
testing.tRunner.func1()
	/usr/lib/golang/src/testing/testing.go:2131 +0x349
panic({0x55858618d050?, 0x1e64514823a8?})
	/usr/lib/golang/src/runtime/panic.go:859 +0x125
reflect.Value.Int(...)
	/usr/lib/golang/src/reflect/value.go:1471
github.com/sibprogrammer/xq/internal/utils.FormatJson({0x5585862029f0?, 0x1e645158dd40?}, {0x558586202950, 0x1e64514e4900}, {0x558585e2c02b, 0x2}, 0x2)
	/builddir/build/BUILD/xq-1.4.0-build/xq-1.4.0/internal/utils/utils.go:446 +0x1107
github.com/sibprogrammer/xq/internal/utils.TestXmlToJSON(0x1e645162e908)
	/builddir/build/BUILD/xq-1.4.0-build/xq-1.4.0/internal/utils/jsonutil_test.go:43 +0x591
testing.tRunner(0x1e645162e908, 0x558586205b98)
	/usr/lib/golang/src/testing/testing.go:2190 +0xea
created by testing.(*T).Run in goroutine 1
	/usr/lib/golang/src/testing/testing.go:2255 +0x4d4
FAIL	github.com/sibprogrammer/xq/internal/utils	0.008s
FAIL
~~~

Go's backward-compatibility promise does not cover unexported fields. In Go 1.27, the internal state machine of encoding/json was refactored. Because tokenState no longer exists (or has changed shape), this reflection hack now causes a panic or compilation failure.

**Steps to Reproduce**

Try with go-1.27rc2

--- a/internal/utils/utils.go
+++ b/internal/utils/utils.go
@@ -9,7 +9,6 @@ import (
 	"io"
 	"os"
 	"os/exec"
-	"reflect"
 	"regexp"
 	"strconv"
 	"strings"
@@ -43,18 +42,6 @@ type QueryOptions struct {
 	Indent   string
 	Colors   int
 }
-
-const (
-	jsonTokenTopValue = iota
-	jsonTokenArrayStart
-	jsonTokenArrayValue
-	jsonTokenArrayComma
-	jsonTokenObjectStart
-	jsonTokenObjectKey
-	jsonTokenObjectColon
-	jsonTokenObjectValue
-	jsonTokenObjectComma
-)
 
 func FormatXml(reader io.Reader, writer io.Writer, indent string, colors int) error {
 	decoder := xml.NewDecoder(reader)
@@ -435,12 +422,125 @@ func FormatJson(reader io.Reader, writer io.Writer, indent string, colors int) e
 	attrColor := color.New(color.FgHiBlue).SprintFunc()
 	valueColor := color.New(color.FgGreen).SprintFunc()
 
-	level := 0
-	suffix := ""
-	prefix := ""
 	newline := "\n"
 	if indent == "" {
 		newline = ""
+	}
+	write := func(args ...any) error {
+		_, err := fmt.Fprint(writer, args...)
+		return err
+	}
+
+	var formatToken func(json.Token, int) error
+	formatToken = func(token json.Token, level int) error {
+		switch typedToken := token.(type) {
+		case json.Delim:
+			switch rune(typedToken) {
+			case '{':
+				if err := write(tagColor("{")); err != nil {
+					return err
+				}
+				level++
+				index := 0
+				for decoder.More() {
+					if index > 0 {
+						if err := write(",", newline); err != nil {
+							return err
+						}
+					} else if err := write(newline); err != nil {
+						return err
+					}
+					if err := write(strings.Repeat(indent, level)); err != nil {
+						return err
+					}
+
+					keyToken, err := decoder.Token()
+					if err != nil {
+						return err
+					}
+					key, ok := keyToken.(string)
+					if !ok {
+						return fmt.Errorf("expected JSON object key, got %T", keyToken)
+					}
+
+					valueToken, err := decoder.Token()
+					if err != nil {
+						return err
+					}
+
+					if err := write(attrColor(strconv.Quote(key)), ": "); err != nil {
+						return err
+					}
+					if err := formatToken(valueToken, level); err != nil {
+						return err
+					}
+					index++
+				}
+				level--
+
+				endToken, err := decoder.Token()
+				if err != nil {
+					return err
+				}
+				if endToken != json.Delim('}') {
+					return fmt.Errorf("expected JSON object end, got %v", endToken)
+				}
+				if index > 0 {
+					return write(newline, strings.Repeat(indent, level), tagColor("}"))
+				}
+				return write(tagColor("}"))
+			case '[':
+				if err := write(tagColor("[")); err != nil {
+					return err
+				}
+				level++
+				index := 0
+				for decoder.More() {
+					if index > 0 {
+						if err := write(",", newline); err != nil {
+							return err
+						}
+					} else if err := write(newline); err != nil {
+						return err
+					}
+					if err := write(strings.Repeat(indent, level)); err != nil {
+						return err
+					}
+
+					valueToken, err := decoder.Token()
+					if err != nil {
+						return err
+					}
+					if err := formatToken(valueToken, level); err != nil {
+						return err
+					}
+					index++
+				}
+				level--
+
+				endToken, err := decoder.Token()
+				if err != nil {
+					return err
+				}
+				if endToken != json.Delim(']') {
+					return fmt.Errorf("expected JSON array end, got %v", endToken)
+				}
+				if index > 0 {
+					return write(newline, strings.Repeat(indent, level), tagColor("]"))
+				}
+				return write(tagColor("]"))
+			default:
+				return fmt.Errorf("unexpected JSON delimiter %q", typedToken)
+			}
+		case string:
+			return write(valueColor(strconv.Quote(typedToken)))
+		case float64, json.Number, bool:
+			return write(valueColor(typedToken))
+		case nil:
+			return write(valueColor("null"))
+		}
+
+		return nil
 	}
 
 	for {
@@ -454,66 +554,12 @@ func FormatJson(reader io.Reader, writer io.Writer, indent string, colors int) e
 			return err
 		}
 
-		v := reflect.ValueOf(*decoder)
-		tokenState := v.FieldByName("tokenState").Int()
-
-		switch tokenType := token.(type) {
-		case json.Delim:
-			switch rune(tokenType) {
-			case '{':
-				_, _ = fmt.Fprint(writer, prefix, tagColor("{"), newline)
-				level++
-				suffix = strings.Repeat(indent, level)
-			case '}':
-				if level > 0 {
-					level--
-				}
-				_, _ = fmt.Fprint(writer, newline, strings.Repeat(indent, level), tagColor("}"))
-				if tokenState == jsonTokenArrayComma {
-					suffix = "," + newline + strings.Repeat(indent, level)
-				}
-			case '[':
-				_, _ = fmt.Fprint(writer, prefix, tagColor("["), newline)
-				level++
-				suffix = strings.Repeat(indent, level)
-			case ']':
-				if level > 0 {
-					level--
-				}
-				_, _ = fmt.Fprint(writer, newline, strings.Repeat(indent, level), tagColor("]"))
-			}
-		case string:
-			escapedToken := strconv.Quote(token.(string))
-			value := valueColor(escapedToken)
-			if tokenState == jsonTokenObjectColon {
-				value = attrColor(escapedToken)
-			}
-			_, _ = fmt.Fprintf(writer, "%s%s", prefix, value)
-		case float64:
-			_, _ = fmt.Fprintf(writer, "%s%v", prefix, valueColor(token))
-		case json.Number:
-			_, _ = fmt.Fprintf(writer, "%s%v", prefix, valueColor(token))
-		case bool:
-			_, _ = fmt.Fprintf(writer, "%s%v", prefix, valueColor(token))
-		case nil:
-			_, _ = fmt.Fprintf(writer, "%s%s", prefix, valueColor("null"))
+		if err := formatToken(token, 0); err != nil {
+			return err
 		}
-
-		switch tokenState {
-		case jsonTokenObjectColon:
-			suffix = ": "
-		case jsonTokenObjectComma:
-			suffix = "," + newline + strings.Repeat(indent, level)
-		case jsonTokenArrayComma:
-			suffix = "," + newline + strings.Repeat(indent, level)
-		}
-
-		prefix = suffix
 	}
 
-	_, _ = fmt.Fprint(writer, "\n")
-
-	return nil
+	return write("\n")
 }
 
 func IsHTML(input string) bool {

--- a/internal/utils/utils_test.go
+++ b/internal/utils/utils_test.go
@@ -2,6 +2,7 @@ package utils
 
 import (
 	"bytes"
+	"errors"
 	"io"
 	"os"
 	"path/filepath"
@@ -10,6 +11,28 @@ import (
 
 	"github.com/stretchr/testify/assert"
 )
+
+var errWriteFailed = errors.New("write failed")
+
+type failingWriter struct{}
+
+func (failingWriter) Write([]byte) (int, error) {
+	return 0, errWriteFailed
+}
+
+type failOnWriteWriter struct {
+	count  int
+	failOn int
+}
+
+func (writer *failOnWriteWriter) Write(data []byte) (int, error) {
+	writer.count++
+	if writer.count == writer.failOn {
+		return 0, errWriteFailed
+	}
+
+	return len(data), nil
+}
 
 func getFileReader(filename string) io.Reader {
 	reader, err := os.Open(filename)
@@ -100,6 +123,32 @@ func TestFormatJson(t *testing.T) {
 		assert.Nil(t, formatErr)
 		assert.Equal(t, expectedJson, output.String())
 	}
+
+	tests := []struct {
+		name     string
+		input    string
+		expected string
+	}{
+		{"empty object", "{}", "{}\n"},
+		{"empty array", "[]", "[]\n"},
+		{"nested empty containers in object", "{\"a\":{},\"b\":[]}", "{\n  \"a\": {},\n  \"b\": []\n}\n"},
+		{"nested empty containers in array", "[{},[]]", "[\n  {},\n  []\n]\n"},
+	}
+
+	for _, testCase := range tests {
+		t.Run(testCase.name, func(t *testing.T) {
+			output := new(strings.Builder)
+			formatErr := FormatJson(strings.NewReader(testCase.input), output, "  ", ColorsDisabled)
+			assert.NoError(t, formatErr)
+			assert.Equal(t, testCase.expected, output.String())
+		})
+	}
+
+	formatErr := FormatJson(strings.NewReader("{}"), failingWriter{}, "  ", ColorsDisabled)
+	assert.ErrorIs(t, formatErr, errWriteFailed)
+
+	formatErr = FormatJson(strings.NewReader("{}"), &failOnWriteWriter{failOn: 3}, "  ", ColorsDisabled)
+	assert.ErrorIs(t, formatErr, errWriteFailed)
 }
 
 func TestXPathQuery(t *testing.T) {

